src/Components/Components/src/Routing/Router.cs

```diff
@@ -194,51 +195,81 @@ internal virtual void Refresh(bool isNavigationIntercepted)
             if (!OnNavigateAsync.HasDelegate)
             {
                 return true;
             }
 
-            // If we've already invoked a task and stored its CTS, then
-            // cancel that existing CTS.
+            // Cancel the CTS instead of disposing it, since disposing does not
+            // actually cancel and can cause unintended Object Disposed Exceptions.
+            // This effectivelly cancels the previously running task and completes it.
             _onNavigateCts?.Cancel();
-            // Then make sure that the task has been completed cancelled or
-            // completed before continuing with the execution of this current task.
+            // Then make sure that the task has been completely cancelled or completed
+            // before starting the next one. This avoid race conditions where the cancellation
+            // for the previous task was set but not fully completed by the time we get to this
+            // invocation.
             await previousOnNavigate;
 
-            // Create a new cancellation token source for this instance
             _onNavigateCts = new CancellationTokenSource();
             var navigateContext = new NavigationContext(path, _onNavigateCts.Token);
 
-            // Create a cancellation task based on the cancellation token
-            // associated with the current running task.
             var cancellationTcs = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
             navigateContext.CancellationToken.Register(state =>
                 ((TaskCompletionSource)state).SetResult(), cancellationTcs);
 
             var task = OnNavigateAsync.InvokeAsync(navigateContext);
 
-            // If the user provided a Navigating render fragment, then show it.
             if (Navigating != null && task.Status != TaskStatus.RanToCompletion)
             {
                 _renderHandle.Render(Navigating);
             }
 
             var completedTask = await Task.WhenAny(task, cancellationTcs.Task);
-            return task == completedTask;
+
+            // We should not render the component associated with the route we
+            // are navigating to if:
+            // - The OnNavigateAsync task threw an exception
+            // - The OnNavigateAsync task wasn't cancelled by our CTS
+            // - The cancellation task completes first
+            if (task == completedTask)
+            {
+                if (completedTask.IsCompletedSuccessfully)
+                {
+                    return true;
+                }
+
+                if (completedTask.IsFaulted)
+                {
+                    ThrowExceptionInRenderer(completedTask);
+                    return false;
+                }
+
+                if (completedTask.IsCanceled)
+                {
+                    throw new InvalidOperationException("OnNavigateAsync callback cannot be canceled.");
+                }
```

Comment:
I'm not sure this is quite right. If `task` was cancelled, it's possible that's just a side-effect of our code triggering the cancellation token. In that case we have a race condition where either `task` or `completedTask` might notify first, but in both cases it should be treated as "we cancelled this task so it's not an error".

So I think the check here should be something like `if (!_onNavigateCts.Token.IsCancellationRequested)`. Additionally, we don't want to throw an `InvalidOperationException` since this is a task cancellation - we'd want a `TaskCancelledException` or `OperationCancelledException`. This is why I was suggesting `await`-ing the `completedTask` so that we get whatever is the right kind of exception automatically.

Reply:
> Additionally, we don't want to throw an InvalidOperationException since this is a task cancellation - we'd want a TaskCancelledException or OperationCancelledException. 

Hm -- I might be misunderstanding but if our intent is to communicate to developers that cancelling the `OnNavigateAsync` task themselves is not acceptable than shouldn't an IOE exception makes more sense?

Reply:
I'll defer to you on this! TBH I don't know for sure what are the most idiomatic patterns around notifying about illegal cancellation. The one thing I had thought was maybe we wanted the exception stack trace to refer to the place where the underlying task was cancelled rather than referring to line 249 here, but I really don't know for sure whether that's what would happen if we replaced L249 with `await completedTask;` to trigger the exception that way.

Reply:
I ended up throwing the IOE exception but setting the inner exception to the originally thrown value.